File: core/src/main/java/com/pluxity/station/StationCodeService.java
package com.pluxity.station;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.CollectionUtils;

@Service
@RequiredArgsConstructor
public class StationCodeService {

    private final StationCodeRepository stationCodeRepository;

    @Transactional
    public void save(Station station, String code) {
        stationCodeRepository.save(StationCode.builder().station(station).code(code).build());
    }

    @Transactional(readOnly = true)
    public List<String> findCodesByStation(Station station) {
        return stationCodeRepository.findByStationOrderByCreatedAtDesc(station).stream()
                .map(StationCode::getCode)
                .toList();
    }

    @Transactional
    public void deleteByStation(Station station) {
        stationCodeRepository.deleteByStation(station);
    }

    @Transactional(readOnly = true)
    public Map<Station, List<String>> findCodeMapByStationIds(List<Station> stations) {
        if (CollectionUtils.isEmpty(stations)) {
            return Collections.emptyMap();
        }
        return stationCodeRepository.findByStationInOrderByCreatedAtDesc(stations).stream()
                .collect(
                        Collectors.groupingBy(
                                StationCode::getStation,
                                Collectors.mapping(StationCode::getCode, Collectors.toList())));
    }
}

File: core/src/main/kotlin/com/pluxity/station/StationCodeService.kt
package com.pluxity.station

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.CollectionUtils

@Service
class StationCodeService(
    private val stationCodeRepository: StationCodeRepository,
) {
    @Transactional
    fun save(
        station: Station,
        code: String,
    ) {
        stationCodeRepository.save(StationCode(station = station, code = code))
    }

    @Transactional(readOnly = true)
    fun findCodesByStation(station: Station): List<String> =
        stationCodeRepository.findByStationOrderByCreatedAtDesc(station).map { it.code }

    @Transactional
    fun deleteByStation(station: Station) {
        stationCodeRepository.deleteByStation(station)
    }

    @Transactional(readOnly = true)
    fun findCodeMapByStationIds(stations: List<Station>): Map<Station, List<String>> {
        if (CollectionUtils.isEmpty(stations)) {
            return mapOf()
        }
        return stationCodeRepository
            .findByStationInOrderByCreatedAtDesc(stations)
            .groupBy { it.station }
            .mapValues { (_, codes) -> codes.map { it.code } }
    }
}
